inc/class-wp-cli-command.php

```diff
@@ -59,94 +58,74 @@ public function assets( $args, $args_assoc ) {
+		$output_dir = $args[0] ?? '';
+
+		$status = null;
+		foreach( $urls as $url ) {
 			$progress->tick();
-			return $contents;
-		}, $urls );
-		$progress->finish();
 
-		if ( ! empty( $args_assoc['replace-from'] ) ) {
-			add_filter( 'static_page_replace_urls_in_content', function( $contents ) use ( $args_assoc ) {
-				return str_replace( $args_assoc['replace-from'], $args_assoc['replace-to'], $contents );
-			});
-		}
+			$pid = \pcntl_fork();
 
-		foreach ( $contents as &$content ) {
-			$content = replace_urls( $content, $args_assoc['config'] );
-		}
+			if ( $pid == - 1 ) {
+				die( 'could not fork' );
+			} else if ( $pid ) {
+				// This is the main process. Put it to wait till all the child process finished.
+				\pcntl_wait( $status ); //Protect against Zombie children
+			} else {
+				// Now we are in child process.
 
-		$progress = WP_CLI\Utils\make_progress_bar( 'Saving pages', count( $urls ) );
+				$content = get_url_contents( $url, $config );
 
-		if ( ! empty( $args[0] ) ) {
-			add_filter( 'static_page_destination_directory', function( $dir ) use ( $args ) {
-				return $args[0];
-			});
-		}
-		array_map( function( $content, $url ) use ( $progress, $args_assoc ) {
-			$progress->tick();
-			if ( ! empty( $args_assoc['verbose'] ) ) {
-				WP_CLI::line( 'Saving ' . $url );
-			}
+				if ( is_wp_error( $content ) ) {
+					WP_CLI::warning( $content->get_error_message() );
+					$content = '';
+				}
 
-			$options = [
-				'user_id' => get_current_user_id(),
-				'context' => '',
-				'action'  => '',
-			];
-
-			$post_id = url_to_postid( $url );
-			if ( ! empty( $post_id ) ) {
-				$post = get_post( $post_id );
-				if ( $post instanceof  \WP_Post && $post->post_type === 'page' ) {
-					$options['context'] = 'page';
-					$options['action']  = 'wp_cli_netstorage_publish';
+				if ( ! empty( $replace_from ) ) {
+					add_filter( 'static_page_replace_urls_in_content', function( $contents ) use ( $replace_from, $replace_to ) {
+						return str_replace( $replace_from, $replace_to, $contents );
+					});
 				}
-			} else {
-				$parsed_url = wp_parse_url( $url );
-				if ( ! empty( $parsed_url ) ) {
-					$args = array(
-						'meta_key'   => 'netstorage_path',
-						'meta_value' => substr( $parsed_url['path'], 1 ),
-						'post_type'  => 'netstorage-file',
-					);
-
-					$query = new \WP_Query( $args );
-					if ( ! empty( $query->post ) && $query->post->ID === $post_id ) { // Get the first post assuming no posts will have the same path.
-						$options['context'] = 'netstorage-file';
-						$options['action']  = 'wp_cli_netstorage_publish';
-					}
+
+				$content = replace_urls( $content, $config );
+
+				if ( ! empty( $output_dir ) ) {
+					add_filter( 'static_page_destination_directory', function( $dir ) use ( $output_dir ) {
+						return $output_dir;
+					});
 				}
+
+				// Fire 3rd party actions.
+				$this->do_actions( $url );
+
+				save_contents_for_url( $content, $url, $config );
```

Comment:
Should we also managed to run this in parallel, it would mean per the above we'd also get _some_ parrallel uploading to netstorage which is nice. It _might_ be technically better to de-couple the capturing output, and saving to netstorage so each can run at maximum page, but for now this should be ok.